reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -580,6 +586,59 @@ def apply_wiener_filter(projection, ctf, small_number):
         projection_wfilter_f = projection * wfilter
         return projection_wfilter_f
 
+    @staticmethod
+    def compute_ssnr(projections, ctfs_list, small_number=0.01):
+        """Compute spectral signal to noise ratio (SSNR) for each pixel of projections.
+
+        Uses section 2.6:
+        https://doi.org/10.1016/j.jsb.2011.06.010
+
+        Parameters
+        ----------
+        projections : arr
+            Shape (n_projections, n_pix, n_pix)
+        ctfs_list : list
+            Shape (n_ctfs, n_pix, n_pix)
+        small_number : float
+            Used for tuning Wiener filter.
+
+        Returns
+        -------
+        ssnr : arr
+            Shape (n_pix, n_pix) the SSNR of each pixel of a projection.
+        """
+        n_pix = len(projections[0])
+        ctfs = np.array(ctfs_list)
+
+        signal_values = np.sum(ctfs * projections, axis=0) / np.sum(
+            ctfs * ctfs + small_number, axis=0
+        )
+
+        ctf_sq_sum = np.zeros(n_pix // 2)
+        ctf_img_sq_sum = np.zeros(n_pix // 2)
+        diff_sq_sum = np.zeros(n_pix // 2)
+        shell_pixels = np.zeros(n_pix // 2)
+
+        for R in range(n_pix // 2):
+            mask = IterativeRefinement.binary_mask(
+                (n_pix // 2, n_pix // 2), R, projections[0].shape, 2
+            )
+            ctf_sq_sum[R] = np.sum(mask * np.sum(ctfs**2, axis=0))
+            ctf_img_sq_sum[R] = np.sum(
+                mask * np.sum(ctfs**2 * np.abs(projections) ** 2, axis=0)
+            )
+            diff_sq_sum[R] = np.sum(
+                mask * np.sum(np.abs(projections - ctfs * signal_values) ** 2, axis=0)
+            )
+            shell_pixels[R] = np.sum(mask)
+
+        sigma_rs_2 = ctf_img_sq_sum / ctf_sq_sum
+        sigma_rn_2 = diff_sq_sum / (shell_pixels * (len(projections) - 1))
+
+        ssnr_1d = (sigma_rs_2 / sigma_rn_2) - shell_pixels / ctf_sq_sum
+
+        return IterativeRefinement.expand_1d_to_Nd(ssnr_1d, d=2)
```

Comment:
I would just return it in 1D, and then write a separate wrapper to make into 2D. That way, it's more modular if it's wanted in 3D. Another alternative is to pass in `d` as an argument to `compute_ssnr` (could use `d=2` as default)